events/testdata/cognito-event-userpools-pretokengen-v2.json

```diff
@@ -0,0 +1,70 @@
+{
+  "version": "2",
+  "triggerSource": "TokenGeneration_Authentication",
+  "region": "us-west-2",
+  "userPoolId": "us-east-1_EXAMPLE",
+  "userName": "brcotter",
+  "callerContext": {
+    "awsSdkVersion": "aws-sdk-unknown-unknown",
+    "clientId": "1example23456789"
+  },
+  "request": {
+    "userAttributes": {
+      "sub": "a36036a8-9061-424d-a737-56d57dae7bc6",
+      "cognito:email_alias": "testuser@example.com",
+      "cognito:user_status": "CONFIRMED",
+      "email_verified": "true",
+      "email": "testuser@example.com"
+    },
+    "groupConfiguration": {
+      "groupsToOverride": [],
+      "iamRolesToOverride": []
+    },
+    "scopes": [
+      "aws.cognito.signin.user.admin"
+    ]
+  },
+  "response": {
+    "claimsAndScopeOverrideDetails": {
+      "idTokenGeneration": {
+        "claimsToAddOrOverride": {
+          "family_name": "xyz"
+        },
+        "claimsToSuppress": [
+          "email",
+          "birthdate"
+        ]
+      },
+      "accessTokenGeneration": {
+        "claimsToAddOrOverride": {
+          "family_name": "xyz"
+        },
+        "claimsToSuppress": [
+          "email",
+          "birthdate"
+        ],
+        "scopesToAdd": [
+          "scope1",
+          "scope2",
+          "scopeLomond"
+        ],
+        "scopesToSuppress": [
+          "phone_number"
+        ]
+      },
+      "groupOverrideDetails": {
+        "groupsToOverride": [
+          "group-A",
+          "group-B",
+          "group-C"
+        ],
+        "iamRolesToOverride": [
+          "arn:aws:iam::123456789012:role/sns_callerA",
+          "arn:aws:iam::123456789012:role/sns_callerB",
+          "arn:aws:iam::123456789012:role/sns_callerC"
+        ],
+        "preferredRole": "arn:aws:iam::123456789012:role/sns_caller"
```

Comment:
ok so this one is missing in the structs per the test output

```
         	            	--- Expected
        	            	+++ Actual
        	            	@@ -6,4 +6,5 @@
        	            	  (string) (len=6) "region": (string) (len=9) "us-west-2",
        	            	- (string) (len=7) "request": (map[string]interface {}) (len=3) {
        	            	-  (string) (len=18) "groupConfiguration": (map[string]interface {}) (len=2) {
        	            	+ (string) (len=7) "request": (map[string]interface {}) (len=4) {
        	            	+  (string) (len=14) "clientMetadata": (interface {}) <nil>,
        	            	+  (string) (len=18) "groupConfiguration": (map[string]interface {}) (len=3) {
        	            	    (string) (len=16) "groupsToOverride": ([]interface {}) {
        	            	@@ -11,3 +12,4 @@
        	            	    (string) (len=18) "iamRolesToOverride": ([]interface {}) {
        	            	-   }
        	            	+   },
        	            	+   (string) (len=13) "preferredRole": (interface {}) <nil>
        	            	   },
```

Reply:
and `clientMetadata` is missing in the sample event, though it's present in the docs and the v1 event

and docs make it look pretty optional


> To pass this data to your Lambda function, use the ClientMetadata parameter in the [AdminRespondToAuthChallenge](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminRespondToAuthChallenge.html) and [RespondToAuthChallenge](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_RespondToAuthChallenge.html) API operations. Amazon Cognito doesn't include data from the ClientMetadata parameter in [AdminInitiateAuth](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminInitiateAuth.html) and [InitiateAuth](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html) API operations in the request that it passes to the pre token generation function.

so an `omitempty` annotation should make the ser-de test happy, and not make the sample data any less correct

Reply:
misread the test output of the `preferredRole`, it's that the struct was re-used in a context that didn't have it populated, and serialization of the struct wrote a `null` instead of omitting the value

need to tweak the test a bit, so things that expect `prefferedRole: null` don't barf

Reply:
Went with adding a `null` entry to the test event instead of leaving if undefined. This lets the marshal/unmarshal test stay in the same style as the other tests.